gapis/api/vulkan/vulkan.go

```diff
@@ -78,6 +78,137 @@ func (s *State) SetupInitialState(ctx context.Context, state *api.GlobalState) {
+	}
+	for _, i := range s.LastDrawInfos().All() {
+		for _, d := range i.DescriptorSets().All() {
+			descriptorSets[d.VulkanHandle()] = struct{}{}
+		}
+		pipelines[i.GraphicsPipeline().VulkanHandle()] = struct{}{}
+	}
+
+	// Record usage in the trace commands
+	// top-level commands
+	postCmdCb := func(s *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd) {
+		switch cmd := cmd.(type) {
+		case *VkFreeDescriptorSets:
+			count := cmd.DescriptorSetCount()
+			ds, err := cmd.PDescriptorSets().Slice(0, (uint64)(count), s.MemoryLayout).Read(ctx, cmd, s, nil)
+			if err != nil {
+				panic("Read error")
+			}
+			for _, d := range ds {
+				descriptorSets[d] = struct{}{}
+			}
+
+		case *VkDestroyPipeline:
+			pipelines[cmd.Pipeline()] = struct{}{}
+		}
+
+	}
+	// sub-commands
+	postSubCmdCb := func(state *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd, i interface{}) {
+		vkState := GetState(state)
+		cmdRef, ok := i.(CommandReferenceʳ)
+		if !ok {
+			panic("In Vulkan, MutateWithSubcommands' postSubCmdCb 'interface{}' is not a CommandReferenceʳ")
+		}
+		cmdArgs := GetCommandArgs(ctx, cmdRef, vkState)
+
+		switch args := cmdArgs.(type) {
+		case VkCmdBindDescriptorSetsArgsʳ:
+			for _, d := range args.DescriptorSets().All() {
+				descriptorSets[d] = struct{}{}
+			}
+
+		case VkCmdBindPipelineArgsʳ:
+			pipelines[args.Pipeline()] = struct{}{}
+		}
+	}
+	c, err := capture.ResolveGraphicsFromPath(ctx, capturePath)
+	if err != nil {
+		return err
+	}
+	if err := sync.MutateWithSubcommands(ctx, capturePath, c.Commands, postCmdCb, nil, postSubCmdCb); err != nil {
+		return err
+	}
+
+	// Transitive dependencies
+
+	// Each pipeline may be derived from a base pipeline, in which case this
+	// base pipeline must be added to the list of used pipelines. Loop on this
+	// until we have a stable number of pipelines.
+	for numPipelines := 0; numPipelines != len(pipelines); {
+		numPipelines = len(pipelines)
+		for p := range pipelines {
+			// For both graphics and compute derivative pipelines which are
+			// created using BasePipelineIndex, our API implementation makes
+			// sure that the relevant pipeline handle is set in BasePipeline.
+			// Thus, we can safely use the value in BasePipeline. See the
+			// post-fence code in vkCreate*Pipelines in
+			// gapis/api/vulkan/api/pipeline.api
+			g := s.GraphicsPipelines().Get(p)
+			if !g.IsNil() && (VkPipelineCreateFlagBits(g.Flags())&VkPipelineCreateFlagBits_VK_PIPELINE_CREATE_DERIVATIVE_BIT) != 0 {
+				pipelines[g.BasePipeline()] = struct{}{}
+			}
+			c := s.ComputePipelines().Get(p)
+			if !c.IsNil() && (VkPipelineCreateFlagBits(c.Flags())&VkPipelineCreateFlagBits_VK_PIPELINE_CREATE_DERIVATIVE_BIT) != 0 {
+				pipelines[c.BasePipeline()] = struct{}{}
+			}
+		}
+	}
+
+	// Remove unused parts.
+	var startSize int
+
+	startSize = s.DescriptorSets().Len()
+	for h := range s.DescriptorSets().All() {
+		if _, ok := descriptorSets[h]; !ok {
```

Comment:
so here the struct that we store is actually not used at all, it is just there because Go doesn't have sets(afaik) Can we just store a boolean or some other plain type?

Reply:
I think `struct{}` is better than `bool` since it is more clear in showing that we don't care about what we store in the map. Storing a bool may mean that you give a semantic to that boolean (i.e. you behave differently depending on that boolean value). With type `struct{}`, you have only one possible value (`struct{}{}`), so the only thing you can really test / care about is presence of the key in the map.